snapcraft/internal/repo/snaps.py

```diff
@@ -0,0 +1,214 @@
+                current_channel = 'latest/stable'
+        return current_channel
+
+    def is_classic(self):
+        store_channels = self._get_store_channels()
+        return store_channels[self.channel]['confinement'] == 'classic'
+
+    def is_valid(self):
+        """Check if the snap is valid."""
+        if not self.on_store:
+            return False
+        store_channels = self._get_store_channels()
+        return self.channel in store_channels.keys()
+
+    def install(self):
+        """Installs the snap onto the system."""
+        snap_install_cmd = []
+        if _snap_command_requires_sudo():
+            snap_install_cmd = ['sudo']
+        snap_install_cmd.extend(['snap', 'install', self.name])
+        if self._original_channel:
+            snap_install_cmd.extend(['--channel', self._original_channel])
+        if self.is_classic:
+            # TODO make this a user explicit choice
+            snap_install_cmd.extend(['--classic'])
+        try:
+            check_call(snap_install_cmd)
+        except CalledProcessError as install_error:
+            raise errors.SnapInstallError(snap_name=self.name,
+                                          snap_channel=self.channel)
+
+    def refresh(self):
+        """Refreshes a snap onto a channel on the system."""
+        snap_refresh_cmd = []
+        if _snap_command_requires_sudo():
+            snap_refresh_cmd = ['sudo']
+        snap_refresh_cmd.extend(['snap', 'refresh', self.name,
+                                 '--channel', self.channel])
+        if self.is_classic:
+            # TODO make this a user explicit choice
+            snap_refresh_cmd.extend(['--classic'])
+        try:
+            check_call(snap_refresh_cmd)
+        except CalledProcessError as install_error:
+            raise errors.SnapRefreshError(snap_name=self.name,
+                                          snap_channel=self.channel)
+
+
+def install_snaps(snaps_list):
+    """Install snaps of the format <snap-name>/<channel>."""
+    for snap in snaps_list:
+        snap_pkg = SnapPackage(snap)
+        if not snap_pkg.installed and snap_pkg.on_store:
+            snap_pkg.install()
+        elif snap_pkg.get_current_channel() != snap_pkg.channel:
+            snap_pkg.refresh()
+
+
+def _snap_command_requires_sudo():
+    # snap whoami returns - if the user is not logged in.
+    output = check_output(['snap', 'whoami'])
+    whoami = output.decode(sys.getfilesystemencoding())
+    requires_root = False
+    try:
+        requires_root = whoami.split(':')[1].strip() == '-'
+    # A safeguard if the output changes
+    except IndexError:
+        requires_root = True
+    if requires_root:
+        logger.warning('snapd is not logged in, snap install '
+                       'commands will use sudo')
+    return requires_root
+
+
+def _get_parsed_snap(snap):
+    if '/' in snap:
+        snap_name = snap[:snap.find('/')]
+        snap_channel = snap[snap.find('/')+1:]
```

Comment:
The two `snap.find('/')`s here are bothering me-- perhaps it should be saved in a variable?

Reply:
ok, fair enough